### DISCUSSION TITLE
cdist implementation

### DISCUSSION THREAD

**Discussion by vishpat** [huggingface/candle#1833] [Q&A] [answered]
Does candle have an implementation of the [cdist](https://pytorch.org/docs/stable/generated/torch.cdist.html) ?

**Comment by LaurentMazare** [comment, marked as answer]
It doesn't but you could implement it manually via the following, though it wouldn't be very efficient.
```rust
let x1 = x1.unsqueeze(1)?;
let x2 = x2.unsqueeze(2)?;
let cdist = x1.broadcast_sub(&x2)?.sqr()?.sum(D::Minus1)?.sqrt()?;
```
If this ends up being an issue performance wise, you could also write a custom op for it, e.g. in encodec we have such a custom kernel [here](https://github.com/huggingface/candle/blob/df5f69444e438a7cd8d8ab4971579bf309b72114/candle-transformers/src/models/encodec.rs#L161) though it applies both a cdist and an argmin in the same kernel.
